## Analyzing data of used cars from eBay

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle. 

The data dictionary provided with data is as follows:

   + dateCrawled - When this ad was first crawled. All field-values are taken from this date.
   + name - Name of the car.
   + seller - Whether the seller is private or a dealer.
   + offerType - The type of listing
   + price - The price on the ad to sell the car.
   + abtest - Whether the listing is included in an A/B test.
   + vehicleType - The vehicle Type.
   + yearOfRegistration - The year in which the car was first registered.
   + gearbox - The transmission type.
   + powerPS - The power of the car in PS.
   + model - The car model name.
   + kilometer - How many kilometers the car has driven.
   + monthOfRegistration - The month in which the car was first registered.
   + fuelType - What type of fuel the car uses.
   + brand - The brand of the car.
   + notRepairedDamage - If the car has a damage which is not yet repaired.
   + dateCreated - The date on which the eBay listing was created.
   + nrOfPictures - The number of pictures in the ad.
   + postalCode - The postal code for the location of the vehicle.
   + lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. 

Let's start by importing the libraries we need and reading the dataset into pandas. 

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

Print information about the autos dataframe, as well as the first few rows. We begin the analysis from column names of the dataframe. 

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

According to the information above, we can make the following observations:
1. The dataset contains 20 columns, and 15 columns are strings.
2. Some of the columns have null values, but the percentages of null values are less than 20%.
3. The column names use camelcase instead of snakecase.

## Cleaning column names

Next, we will convert the column name from camelcase to snakecase, and rewrite some of the column names to be more descriptive.

In [3]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [4]:
autos.columns = ["date_crawled", "name", "seller", "offer_type", "price", "abtest", 
                "vehicle_type", "registration_year", "gearbox", "power_ps", "model",
                "odometer", "registration_month", "fuel_type", "brand", "unrepaired_damage",
                "ad_created", "nr_pictures", "postal_code", "last_seen"]

Let's look at the current state of the autos dataframe.

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial exploration and cleaning

Next we will carry out some basic data exploration to determine what other cleaning tasks need to be done. First we need to take a look at descriptive statistics for all columns.

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In [7]:
print(autos["nr_pictures"].value_counts())

0    50000
Name: nr_pictures, dtype: int64


In [8]:
print(autos['seller'].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64


In [9]:
print(autos['offer_type'].value_counts())

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


Now we know there is only one value for column "nr_pictures", two unique values for both "seller" and "offer_type" columns. We need to drop them.

In [10]:
autos = autos.drop(['nr_pictures','seller', 'offer_type'], axis = 1)

The minimum value for columns "registration_year", "power_ps" and "registration_month" is 0, which is inaccurate. We need more investigation with these columns later.

There is non-numeric data stored in columns "price" and "odometer". We need to convert those non-numeric characters to a numeric type.

In [11]:
print(autos["price"].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [12]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos.rename({"price":"price_usd"}, axis = 1, inplace = True)
autos["price_usd"] = autos["price_usd"].astype(int)
autos.describe(include = 'all')

date_crawled         name     price_usd abtest vehicle_type  \
count                 50000        50000  5.000000e+04  50000        44905   
unique                48213        38754           NaN      2            8   
top     2016-03-23 19:38:20  Ford_Fiesta           NaN   test    limousine   
freq                      3           78           NaN  25756        12859   
mean                    NaN          NaN  9.840044e+03    NaN          NaN   
std                     NaN          NaN  4.811044e+05    NaN          NaN   
min                     NaN          NaN  0.000000e+00    NaN          NaN   
25%                     NaN          NaN  1.100000e+03    NaN          NaN   
50%                     NaN          NaN  2.950000e+03    NaN          NaN   
75%                     NaN          NaN  7.200000e+03    NaN          NaN   
max                     NaN          NaN  1.000000e+08    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

We will convert the "odometer" column using the same method.

In [13]:
print(autos['odometer'].unique())

['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [14]:
autos['odometer'] = autos['odometer'].str.replace("km", "")
autos['odometer'] = autos['odometer'].str.replace(",", "")
autos['odometer'] = autos['odometer'].astype(int)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)
print(autos['odometer_km'].unique())
autos.describe(include = 'all')

[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]


date_crawled         name     price_usd abtest vehicle_type  \
count                 50000        50000  5.000000e+04  50000        44905   
unique                48213        38754           NaN      2            8   
top     2016-03-23 19:38:20  Ford_Fiesta           NaN   test    limousine   
freq                      3           78           NaN  25756        12859   
mean                    NaN          NaN  9.840044e+03    NaN          NaN   
std                     NaN          NaN  4.811044e+05    NaN          NaN   
min                     NaN          NaN  0.000000e+00    NaN          NaN   
25%                     NaN          NaN  1.100000e+03    NaN          NaN   
50%                     NaN          NaN  2.950000e+03    NaN          NaN   
75%                     NaN          NaN  7.200000e+03    NaN          NaN   
max                     NaN          NaN  1.000000e+08    NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        50000.000000    47320  50000.000000  47242   50000.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36993           NaN   4024            NaN   
mean          2005.073280      NaN    116.355920    NaN  125732.700000   
std            105.712813      NaN    209.216627    NaN   40042.211706   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     70.000000    NaN  125000.000000   
50%           2003.000000      NaN    105.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

## Exploring the price and odometer columns

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. We will analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [15]:
autos['price_usd'].unique().shape

(2357,)

There are 2357 unique values in 'price_usd' column. Let's take a look at their descriptive statistics. 

In [16]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [17]:
autos['price_usd'].value_counts()

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price_usd, Length: 2357, dtype: int64

We can sort the result to explore the data in detail.

In [18]:
autos['price_usd'].value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_usd, dtype: int64

There are 1421 cars, about 3% of all the cars listed as $0, which is unrealistic. We will remove all them later.

In [19]:
autos['price_usd'].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_usd, dtype: int64

About 14 cars' listing pirce are more than 1 million. According to the sorting result, the price increases steadily before $350000. Since eBay is an auction site, we believe that the cars with price more than 1 million can be removed.

In [20]:
autos = autos[autos["price_usd"].between(1, 351000)]

In [21]:
autos.describe()

price_usd  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

In [22]:
autos["price_usd"].value_counts().sort_index(ascending = True).head(20)

1     156
2       3
3       1
5       2
8       1
9       1
10      7
11      2
12      3
13      2
14      1
15      2
17      3
18      1
20      4
25      5
29      1
30      7
35      1
40      6
Name: price_usd, dtype: int64

In [23]:
autos["price_usd"].value_counts().sort_index(ascending = False).head(20)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
194000    1
190000    1
180000    1
175000    1
169999    1
169000    1
163991    1
163500    1
155000    1
151990    1
Name: price_usd, dtype: int64

Next we will check the "odometer" column.

In [24]:
autos['odometer_km'].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

We can see here that all the values of odometer are rounded. This is probably because there are range options, not exact milage, when sellers fill the information. The majority of the cars have larger odometers based on the value counts result.

## Exporing the date columns

Next we will move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. The "date_crawled", "last_seen", and "ad_created" columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation then we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [25]:
autos.describe(include = 'all')

date_crawled         name      price_usd abtest vehicle_type  \
count                 48565        48565   48565.000000  48565        43979   
unique                46882        37470            NaN      2            8   
top     2016-04-04 16:40:33  Ford_Fiesta            NaN   test    limousine   
freq                      3           76            NaN  25019        12598   
mean                    NaN          NaN    5888.935591    NaN          NaN   
std                     NaN          NaN    9059.854754    NaN          NaN   
min                     NaN          NaN       1.000000    NaN          NaN   
25%                     NaN          NaN    1200.000000    NaN          NaN   
50%                     NaN          NaN    3000.000000    NaN          NaN   
75%                     NaN          NaN    7490.000000    NaN          NaN   
max                     NaN          NaN  350000.000000    NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        48565.000000    46222  48565.000000  46107   48565.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36102           NaN   3900            NaN   
mean          2004.755421      NaN    117.197158    NaN  125770.101925   
std             88.643887      NaN    200.649618    NaN   39788.636804   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     71.000000    NaN  125000.000000   
50%           2004.000000      NaN    107.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         48565.000000     44535       48565             39464   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     29368       10336             34775   
mean              5.782251       NaN         NaN               NaN   
std               3.685595       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 48565  48565.000000                48565  
unique                   76           NaN                38474  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1887           NaN                    8  
mean                    NaN  50975.745207                  NaN  
std                     NaN  25746.968398                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30657.000000                  NaN  
50%                     NaN  49716.000000                  NaN  
75%                     NaN  71665.000000                  NaN  
max                     NaN  99998.000000                  NaN

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values.

In [26]:
autos[["date_crawled", "last_seen", "ad_created"]][:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

You'll notice that the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.
To select the first 10 characters in each column, we can use Series.str[:10].

In [27]:
print(autos["date_crawled"].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
5        2016-03-21
6        2016-03-20
7        2016-03-16
8        2016-03-22
9        2016-03-16
10       2016-03-15
11       2016-03-16
12       2016-03-31
13       2016-03-23
14       2016-03-23
15       2016-04-01
16       2016-03-16
17       2016-03-29
18       2016-03-26
19       2016-03-17
20       2016-03-05
21       2016-03-06
22       2016-03-28
23       2016-03-10
24       2016-04-03
25       2016-03-21
26       2016-04-03
28       2016-03-19
29       2016-04-02
30       2016-03-14
            ...    
49968    2016-04-01
49969    2016-03-17
49970    2016-03-21
49971    2016-03-29
49972    2016-03-26
49973    2016-03-27
49975    2016-03-27
49976    2016-03-19
49977    2016-03-31
49978    2016-04-04
49979    2016-03-20
49980    2016-03-12
49981    2016-03-15
49982    2016-03-29
49983    2016-03-06
49985    2016-04-02
49986    2016-04-04
49987    2016-03-22
49988    2016-03-28


We now will calculate the distribution of values in all of the 3 string columns as percentages. Here we will include missing values in the distribution and to use percentages instead of counts.

In [28]:
autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

To rank by date in ascending order (earliest to latest), chain the Series.sort_index() method.

In [29]:
[autos["date_crawled"].
                       str[:10].
                       value_counts(normalize = True, dropna = False).
                       sort_index(ascending = True)]

[2016-03-05    0.025327
 2016-03-06    0.014043
 2016-03-07    0.036014
 2016-03-08    0.033296
 2016-03-09    0.033090
 2016-03-10    0.032184
 2016-03-11    0.032575
 2016-03-12    0.036920
 2016-03-13    0.015670
 2016-03-14    0.036549
 2016-03-15    0.034284
 2016-03-16    0.029610
 2016-03-17    0.031628
 2016-03-18    0.012911
 2016-03-19    0.034778
 2016-03-20    0.037887
 2016-03-21    0.037373
 2016-03-22    0.032987
 2016-03-23    0.032225
 2016-03-24    0.029342
 2016-03-25    0.031607
 2016-03-26    0.032204
 2016-03-27    0.031092
 2016-03-28    0.034860
 2016-03-29    0.034099
 2016-03-30    0.033687
 2016-03-31    0.031834
 2016-04-01    0.033687
 2016-04-02    0.035478
 2016-04-03    0.038608
 2016-04-04    0.036487
 2016-04-05    0.013096
 2016-04-06    0.003171
 2016-04-07    0.001400
 Name: date_crawled, dtype: float64]

In [30]:
[autos["date_crawled"].
                       str[:10].
                       value_counts(normalize = True, dropna = False).
                       sort_values(ascending = True)]

[2016-04-07    0.001400
 2016-04-06    0.003171
 2016-03-18    0.012911
 2016-04-05    0.013096
 2016-03-06    0.014043
 2016-03-13    0.015670
 2016-03-05    0.025327
 2016-03-24    0.029342
 2016-03-16    0.029610
 2016-03-27    0.031092
 2016-03-25    0.031607
 2016-03-17    0.031628
 2016-03-31    0.031834
 2016-03-10    0.032184
 2016-03-26    0.032204
 2016-03-23    0.032225
 2016-03-11    0.032575
 2016-03-22    0.032987
 2016-03-09    0.033090
 2016-03-08    0.033296
 2016-03-30    0.033687
 2016-04-01    0.033687
 2016-03-29    0.034099
 2016-03-15    0.034284
 2016-03-19    0.034778
 2016-03-28    0.034860
 2016-04-02    0.035478
 2016-03-07    0.036014
 2016-04-04    0.036487
 2016-03-14    0.036549
 2016-03-12    0.036920
 2016-03-21    0.037373
 2016-03-20    0.037887
 2016-04-03    0.038608
 Name: date_crawled, dtype: float64]

The results showed that all the information was crawled evenly between March 5th, 2016 and April 4th, 2016. The data crawled from April 5th, 2016 to April 7th, 2016 was less than before.

With the same method, we will analyze the "last_seen", and "ad_created" columns.

In [31]:
[autos['last_seen'].
                str[:10].
                value_counts(normalize = True, dropna = False).
                sort_index(ascending = True)]

[2016-03-05    0.001071
 2016-03-06    0.004324
 2016-03-07    0.005395
 2016-03-08    0.007413
 2016-03-09    0.009595
 2016-03-10    0.010666
 2016-03-11    0.012375
 2016-03-12    0.023783
 2016-03-13    0.008895
 2016-03-14    0.012602
 2016-03-15    0.015876
 2016-03-16    0.016452
 2016-03-17    0.028086
 2016-03-18    0.007351
 2016-03-19    0.015834
 2016-03-20    0.020653
 2016-03-21    0.020632
 2016-03-22    0.021373
 2016-03-23    0.018532
 2016-03-24    0.019767
 2016-03-25    0.019211
 2016-03-26    0.016802
 2016-03-27    0.015649
 2016-03-28    0.020859
 2016-03-29    0.022341
 2016-03-30    0.024771
 2016-03-31    0.023783
 2016-04-01    0.022794
 2016-04-02    0.024915
 2016-04-03    0.025203
 2016-04-04    0.024483
 2016-04-05    0.124761
 2016-04-06    0.221806
 2016-04-07    0.131947
 Name: last_seen, dtype: float64]

In [32]:
[autos['last_seen'].
                str[:10].
                value_counts(normalize = True, dropna = False).
                sort_values(ascending = True)]

[2016-03-05    0.001071
 2016-03-06    0.004324
 2016-03-07    0.005395
 2016-03-18    0.007351
 2016-03-08    0.007413
 2016-03-13    0.008895
 2016-03-09    0.009595
 2016-03-10    0.010666
 2016-03-11    0.012375
 2016-03-14    0.012602
 2016-03-27    0.015649
 2016-03-19    0.015834
 2016-03-15    0.015876
 2016-03-16    0.016452
 2016-03-26    0.016802
 2016-03-23    0.018532
 2016-03-25    0.019211
 2016-03-24    0.019767
 2016-03-21    0.020632
 2016-03-20    0.020653
 2016-03-28    0.020859
 2016-03-22    0.021373
 2016-03-29    0.022341
 2016-04-01    0.022794
 2016-03-12    0.023783
 2016-03-31    0.023783
 2016-04-04    0.024483
 2016-03-30    0.024771
 2016-04-02    0.024915
 2016-04-03    0.025203
 2016-03-17    0.028086
 2016-04-05    0.124761
 2016-04-07    0.131947
 2016-04-06    0.221806
 Name: last_seen, dtype: float64]

In contrast with "date_crawled" column, majority (~47%) of the listing were last seen between April 5th, 2016 and April 7th, 2016. However, we don't have any clue that whether those cars were selled, or those listing ads were removed by eBay.

In [33]:
[autos['ad_created'].
                     str[:10].
                     value_counts(normalize = True, dropna = False).
                     sort_index(ascending = True)]

[2015-06-11    0.000021
 2015-08-10    0.000021
 2015-09-09    0.000021
 2015-11-10    0.000021
 2015-12-05    0.000021
 2015-12-30    0.000021
 2016-01-03    0.000021
 2016-01-07    0.000021
 2016-01-10    0.000041
 2016-01-13    0.000021
 2016-01-14    0.000021
 2016-01-16    0.000021
 2016-01-22    0.000021
 2016-01-27    0.000062
 2016-01-29    0.000021
 2016-02-01    0.000021
 2016-02-02    0.000041
 2016-02-05    0.000041
 2016-02-07    0.000021
 2016-02-08    0.000021
 2016-02-09    0.000021
 2016-02-11    0.000021
 2016-02-12    0.000041
 2016-02-14    0.000041
 2016-02-16    0.000021
 2016-02-17    0.000021
 2016-02-18    0.000041
 2016-02-19    0.000062
 2016-02-20    0.000041
 2016-02-21    0.000062
                 ...   
 2016-03-09    0.033151
 2016-03-10    0.031895
 2016-03-11    0.032904
 2016-03-12    0.036755
 2016-03-13    0.017008
 2016-03-14    0.035190
 2016-03-15    0.034016
 2016-03-16    0.030125
 2016-03-17    0.031278
 2016-03-18    0.013590
 2016-03-19    0

We notice from the above result, most of the ads were created between the beginning of March 2016 and the beginning of April 2016. The oldest ads was about 10 months older than the newest one.

Let's explore numeric date data, like "registration_year".

In [34]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year when a car was first registered referred to the car's age. There are some odd values here. The minimum was 1000, which is even earlier than the time when cars were invented. The maximum was 9999, which was far in the future. Next, we will deal with these incorrect values.

Our data was crawled in 2016. Because a car can't be first registered after the listing was seen, any vehicle with a registration year later than 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.


In [35]:
[autos[autos['registration_year'].between(1900,2016)]['registration_year'].
                                            value_counts(normalize = True).
                                            sort_index(ascending = True)]

[1910    0.000107
 1927    0.000021
 1929    0.000021
 1931    0.000021
 1934    0.000043
 1937    0.000086
 1938    0.000021
 1939    0.000021
 1941    0.000043
 1943    0.000021
 1948    0.000021
 1950    0.000064
 1951    0.000043
 1952    0.000021
 1953    0.000021
 1954    0.000043
 1955    0.000043
 1956    0.000086
 1957    0.000043
 1958    0.000086
 1959    0.000129
 1960    0.000493
 1961    0.000129
 1962    0.000086
 1963    0.000171
 1964    0.000257
 1965    0.000364
 1966    0.000471
 1967    0.000557
 1968    0.000557
           ...   
 1987    0.001542
 1988    0.002892
 1989    0.003727
 1990    0.007433
 1991    0.007262
 1992    0.007926
 1993    0.009104
 1994    0.013474
 1995    0.026285
 1996    0.029412
 1997    0.041794
 1998    0.050620
 1999    0.062060
 2000    0.067608
 2001    0.056468
 2002    0.053255
 2003    0.057818
 2004    0.057904
 2005    0.062895
 2006    0.057197
 2007    0.048778
 2008    0.047450
 2009    0.044665
 2010    0.034040
 2011    0

In [36]:
(~autos['registration_year'].between(1900, 2016)).sum()/autos.shape[0]

0.038793369710697

The percentages of the cars registered before 1990 were very small. Fewer cars were registered before 1958. That means it is reasonable to remove those  
cars registered before 1900.

In [37]:
#Remove the rows with registration year earlier than 1900 or later than 2016.
autos = autos[autos['registration_year'].between(1900, 2016)]
#Check the registration year in the new dataframe.
autos.describe()

price_usd  registration_year      power_ps    odometer_km  \
count   46681.000000       46681.000000  46681.000000   46681.000000   
mean     5977.716801        2002.910756    117.892933  125586.855466   
std      9177.909479           7.185103    184.922911   39852.528628   
min         1.000000        1910.000000      0.000000    5000.000000   
25%      1250.000000        1999.000000     75.000000  100000.000000   
50%      3100.000000        2003.000000    109.000000  150000.000000   
75%      7500.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        2016.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        46681.000000  46681.000000  
mean             5.827125  51097.434181  
std              3.670300  25755.387192  
min              0.000000   1067.000000  
25%              3.000000  30827.000000  
50%              6.000000  49828.000000  
75%              9.000000  71732.000000  
max             12.000000  99998.000000

With the same method, we will clean "registration_month" column.

In [38]:
#Check the descriptive information of "registration_month" column.
[autos['registration_month'].
                     value_counts(normalize = True, dropna = False).
                     sort_index(ascending = True)]

[0     0.086245
 1     0.066515
 2     0.061053
 3     0.103640
 4     0.083417
 5     0.083053
 6     0.088237
 7     0.079733
 8     0.064716
 9     0.069472
 10    0.074870
 11    0.069172
 12    0.069879
 Name: registration_month, dtype: float64]

Cars were about registered evenly in each month. 8% of the data has 0 as register month. We will keep these data, and change the value of registration month into NaN.

In [39]:
autos.loc[autos['registration_month']==0, 'registration_month'] = np.nan

In [40]:
#Check the descriptive information of "registration_month" column again.
[autos['registration_month'].
                     value_counts(normalize = True, dropna = False).
                     sort_index(ascending = True)]

[ 1.0     0.066515
  2.0     0.061053
  3.0     0.103640
  4.0     0.083417
  5.0     0.083053
  6.0     0.088237
  7.0     0.079733
  8.0     0.064716
  9.0     0.069472
  10.0    0.074870
  11.0    0.069172
  12.0    0.069879
 NaN      0.086245
 Name: registration_month, dtype: float64]

## Exploring price by brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

Let's explore the unique values in the brand column, and decide on which brands we want to aggregate by.

In [41]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [42]:
autos['brand'].value_counts(normalize = True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We are interested in the average prices of the top 20 brands. Now we will aggregate on the top 20 brands. 

In [43]:
#Creat a list of the top 20 brands we will work on.
brands = []
brands = autos['brand'].value_counts(normalize = True).head(20).index
print(brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')


In [44]:
#Loop through the brands list, and calculate average prices for each brand.
avg_price_by_brand = {}
for b in brands:
    selected_rows = autos[autos['brand']==b]
    average = int(selected_rows['price_usd'].mean())
    avg_price_by_brand[b] = average
print(avg_price_by_brand)

{'volkswagen': 5402, 'mini': 10613, 'toyota': 5167, 'opel': 2975, 'ford': 3749, 'mazda': 4112, 'audi': 9336, 'fiat': 2813, 'smart': 3580, 'bmw': 8332, 'sonstige_autos': 12338, 'seat': 4397, 'peugeot': 3094, 'citroen': 3779, 'mercedes_benz': 8628, 'skoda': 6368, 'nissan': 4743, 'volvo': 4946, 'renault': 2474, 'hyundai': 5365}


Here are several observations we can make from the results:

   + Sonstige_autos, Mini, Audi, Mercedes_benz and BMW are the most expensive cars with prices more than $8000;
    
   + Renault, Fiat, Opel and Peugeot are the cheapest cars with prices around $3000;
    
   + Most popular brands, like Volkswagen, Nissan, Volvo, Mazda, Hyundai, and Toyota are selled at prices around $5000.

Another aspect to consider when we purchase a used car is milage. So next we will calculate the average milages of each brand in our brand list.

In [45]:
#Loop through the brands list, and calculate average milage for each brand.
avg_milage_by_brand = {}
for b in brands:
    selected_rows = autos[autos['brand']==b]
    average = int(selected_rows['odometer_km'].mean())
    avg_milage_by_brand[b] = average
print(avg_milage_by_brand)

{'volkswagen': 128707, 'mini': 88105, 'toyota': 115944, 'opel': 129310, 'ford': 124266, 'mazda': 124464, 'audi': 129157, 'fiat': 117121, 'smart': 99326, 'bmw': 132572, 'sonstige_autos': 89956, 'seat': 121131, 'peugeot': 127153, 'citroen': 119694, 'mercedes_benz': 130788, 'skoda': 110848, 'nissan': 118330, 'volvo': 138067, 'renault': 128071, 'hyundai': 106442}


## Storing aggregate data in a new dataframe

For the top 20 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations. It's difficult to compare more than two aggregate series objects if we want to extend to more columns. we can't compare more than a few rows from each series object. 

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly.

In [46]:
#Convert both dictionaries to series objects, using the series constructor.
apb_series = pd.Series(avg_price_by_brand)
amb_series = pd.Series(avg_milage_by_brand)
print(apb_series)
print(amb_series)

audi               9336
bmw                8332
citroen            3779
fiat               2813
ford               3749
hyundai            5365
mazda              4112
mercedes_benz      8628
mini              10613
nissan             4743
opel               2975
peugeot            3094
renault            2474
seat               4397
skoda              6368
smart              3580
sonstige_autos    12338
toyota             5167
volkswagen         5402
volvo              4946
dtype: int64
audi              129157
bmw               132572
citroen           119694
fiat              117121
ford              124266
hyundai           106442
mazda             124464
mercedes_benz     130788
mini               88105
nissan            118330
opel              129310
peugeot           127153
renault           128071
seat              121131
skoda             110848
smart              99326
sonstige_autos     89956
toyota            115944
volkswagen        128707
volvo             138067
dtype: 

In [47]:
#Create a dataframe from the first series object using the dataframe constructor.
avg_price_milage_brand = pd.DataFrame(apb_series, columns = ['avg_price'])
print(avg_price_milage_brand)

                avg_price
audi                 9336
bmw                  8332
citroen              3779
fiat                 2813
ford                 3749
hyundai              5365
mazda                4112
mercedes_benz        8628
mini                10613
nissan               4743
opel                 2975
peugeot              3094
renault              2474
seat                 4397
skoda                6368
smart                3580
sonstige_autos      12338
toyota               5167
volkswagen           5402
volvo                4946


In [48]:
#Assign the other series as a new column in this dataframe.
avg_price_milage_brand['avg_milage'] = amb_series
print(avg_price_milage_brand)

                avg_price  avg_milage
audi                 9336      129157
bmw                  8332      132572
citroen              3779      119694
fiat                 2813      117121
ford                 3749      124266
hyundai              5365      106442
mazda                4112      124464
mercedes_benz        8628      130788
mini                10613       88105
nissan               4743      118330
opel                 2975      129310
peugeot              3094      127153
renault              2474      128071
seat                 4397      121131
skoda                6368      110848
smart                3580       99326
sonstige_autos      12338       89956
toyota               5167      115944
volkswagen           5402      128707
volvo                4946      138067


After building such a dataframe, we can compare different brands from different aspects. Among the cars in the same level, there is a trend that  the cars with the less milages is more expensive.